src/coordinator/tsdb/remote/compressed_codecs.go

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package remote
+
+import (
+	"io"
+	"time"
+
+	"github.com/m3db/m3db/src/coordinator/generated/proto/rpc"
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/encoding/m3tsz"
+	"github.com/m3db/m3db/src/dbnode/ts"
+	"github.com/m3db/m3db/src/dbnode/x/xio"
+	"github.com/m3db/m3db/src/dbnode/x/xpool"
+	"github.com/m3db/m3x/checked"
+	"github.com/m3db/m3x/ident"
+	xtime "github.com/m3db/m3x/time"
+)
+
+func blockReaderToSegment(br xio.BlockReader) (*rpc.Segment, error) {
+	segment, err := br.Segment()
+	if err != nil {
+		return nil, err
+	}
+	return &rpc.Segment{
+		Head: segment.Head.Bytes(),
+		Tail: segment.Tail.Bytes(),
+	}, nil
+}
+
+// RPCFromSeriesIterator converts a seriesIterator to a grpc series
+func RPCFromSeriesIterator(it encoding.SeriesIterator) (*rpc.Series, error) {
+	replicas := it.Replicas()
+	compressedReplicas := make([]*rpc.CompressedValuesReplica, 0, len(replicas))
+	for _, replica := range replicas {
+		replicaSegments := make([]*rpc.Segments, 0, len(replicas))
+		readers := replica.Readers()
+		for next := true; next; next = readers.Next() {
+			segments := &rpc.Segments{}
+			l, _, _ := readers.CurrentReaders()
+			// NB(arnikola) If there's only a single reader, the segment has been merged
+			// otherwise, multiple unmerged segments exist.
+			if l == 1 {
```

Comment:
extract this out into a function ? The func takes readers as input and returns a slice of segments. Also, its an ugly detail for coordinator to know about, any chance we can push this to within m3db ?